Undefined reference to ... when running function from struct that implement interface concurrently

<!-- Please make sure to run `v up` before reporting any issues as it may have already been fixed.
     It's also advisable to update all relevant modules using `v outdated` and `v install` -->

<!-- You can use `v doctor` to fill up the next fields -->
**V version:** V 0.2.2 acf9d16.7ec55e4
**OS:** linux, Linux version 5.4.72-microsoft-standard-WSL2 (oe-user@oe-host) (gcc version 8.2.0 (GCC)) #1 SMP Wed Oct 28 23:40:43 UTC 2020 (WSL 2)

**What did you do?**

Ran this code

```v
interface CanPerformTask {
  task()
}

struct Task1 {}

fn (task1 Task1) task() {}

struct Task2 {}

fn (task2 Task2) task() {}

fn main() {
  mut tasks := []CanPerformTask{}

  tasks << Task1{}
  tasks << Task2{}

  for task in tasks {
    go task.task()
  }
}
```

**What did you expect to see?**

No error on console.

 
**What did you see instead?**

```bash
/tmp/v/test.1005538350077206667.tmp.c: In function 'main__CanPerformTask_task_thread_wrapper':
/tmp/v/test.1005538350077206667.tmp.c:1766:2: warning: implicit declaration of function 'main__CanPerformTask_task' [-Wimplicit-function-declaration]
 1766 |  main__CanPerformTask_task(arg->arg0);
      |  ^~~~~~~~~~~~~~~~~~~~~~~~~
/usr/lib/gcc/x86_64-alpine-linux-musl/10.2.1/../../../../x86_64-alpine-linux-musl/bin/ld: /tmp/ccfAhmjn.o: in function `main__CanPerformTask_task_thread_wrapper':
/tmp/v/test.1005538350077206667.tmp.c:1766: undefined reference to `main__CanPerformTask_task'
collect2: error: ld returned 1 exit status
builder error:
==================
```

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -5935,12 +5935,29 @@ fn (mut g Gen) go_expr(node ast.GoExpr) {
 		} else {
 			g.gowrappers.write_string('\t')
 		}
-		g.gowrappers.write_string('${name}(')
 		if expr.is_method {
-			g.gowrappers.write_string('arg->arg0')
+			unwrapped_rec_type := g.unwrap_generic(expr.receiver_type)
+			typ_sym := g.table.get_type_symbol(unwrapped_rec_type)
+			rec_cc_type := g.cc_type(unwrapped_rec_type, false)
+			receiver_type_name := util.no_dots(rec_cc_type)
+			if typ_sym.kind == .interface_
+				&& (typ_sym.info as ast.Interface).defines_method(expr.name) {
+				g.gowrappers.write_string('${c_name(receiver_type_name)}_name_table[')
+				g.gowrappers.write_string('arg->arg0')
+				dot := if expr.left_type.is_ptr() { '->' } else { '.' }
+				mname := c_name(expr.name)
+				g.gowrappers.write_string('${dot}_typ]._method_${mname}(')
+				g.gowrappers.write_string('arg->arg0')
+				g.gowrappers.write_string('${dot}_object')
+			} else {
+				g.gowrappers.write_string('${name}(')
+				g.gowrappers.write_string('arg->arg0')
+			}
 			if expr.args.len > 0 {
 				g.gowrappers.write_string(', ')
 			}
+		} else {
+			g.gowrappers.write_string('${name}(')
 		}
 		if expr.args.len > 0 {
 			mut has_cast := false

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -5938,10 +5938,10 @@ fn (mut g Gen) go_expr(node ast.GoExpr) {
 		if expr.is_method {
 			unwrapped_rec_type := g.unwrap_generic(expr.receiver_type)
 			typ_sym := g.table.get_type_symbol(unwrapped_rec_type)
-			rec_cc_type := g.cc_type(unwrapped_rec_type, false)
-			receiver_type_name := util.no_dots(rec_cc_type)
 			if typ_sym.kind == .interface_
 				&& (typ_sym.info as ast.Interface).defines_method(expr.name) {
+				rec_cc_type := g.cc_type(unwrapped_rec_type, false)
+				receiver_type_name := util.no_dots(rec_cc_type)
 				g.gowrappers.write_string('${c_name(receiver_type_name)}_name_table[')
 				g.gowrappers.write_string('arg->arg0')
 				dot := if expr.left_type.is_ptr() { '->' } else { '.' }

--- a/vlib/v/tests/go_call_interface_method_test.v
+++ b/vlib/v/tests/go_call_interface_method_test.v
@@ -1,0 +1,28 @@
+interface CanPerformTask {
+	task()
+}
+
+struct Task1 {}
+
+fn (task1 Task1) task() {
+	println('task1')
+}
+
+struct Task2 {}
+
+fn (task2 Task2) task() {
+	println('task2')
+}
+
+fn test_go_call_interface_method() {
+	mut tasks := []CanPerformTask{}
+
+	tasks << Task1{}
+	tasks << Task2{}
+
+	for task in tasks {
+		go task.task()
+	}
+
+	assert true
+}

